streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2002,58 +2002,40 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
   final def mergeWith[R1 <: R, E1 >: E, O2, O3](
     that: ZStream[R1, E1, O2],
     strategy: TerminationStrategy = TerminationStrategy.Both
-  )(l: O => O3, r: O2 => O3): ZStream[R1, E1, O3] = {
-    type Candidate[E, O] = Fiber[Nothing, Take[E, O]]
-    type Loser           = Either[Candidate[E, O], Candidate[E1, O2]]
-
-    def race(left: URIO[R, Take[E, O]], right: URIO[R1, Take[E1, O2]]): URIO[R1, (Take[E1, O3], Loser)] =
-      left.raceWith[R1, Nothing, Nothing, Take[E1, O2], (Take[E1, O3], Loser)](right)(
-        (exit, right) => ZIO.done(exit).map(a => (a.map(_.map(l)), Right(right))),
-        (exit, left) => ZIO.done(exit).map(b => (b.map(_.map(r)), Left(left)))
-      )
-
-    def termination[E, O](left: Boolean, right: Boolean, mapper: O => O3)(fiber: Candidate[E, O]) =
-      strategy match {
-        case TerminationStrategy.Either =>
-          fiber.interrupt.as(Take.End)
-        case TerminationStrategy.Left if left =>
-          fiber.interrupt.as(Take.End)
-        case TerminationStrategy.Right if right =>
-          fiber.interrupt.as(Take.End)
-        case _ =>
-          fiber.join.map(_.map(_.map(mapper))).map(_.map((_, (left, right, None))))
-      }
+  )(l: O => O3, r: O2 => O3): ZStream[R1, E1, O3] =
+    ZStream {
+      import TerminationStrategy.{ Left => L, Right => R, Either => E }
 
-    combineChunks(that)((false, false, Option.empty[Loser])) {
-      case ((leftDone, rightDone, loser), left, right) =>
-        if (leftDone) {
-          right.map(c => (c.map(r), (leftDone, rightDone, None))).run
-        } else if (rightDone) {
-          left.map(c => (c.map(l), (leftDone, rightDone, None))).run
-        } else {
-          val result = loser match {
-            case None               => race(left.run, right.run)
-            case Some(Left(loser))  => race(loser.join, right.run)
-            case Some(Right(loser)) => race(left.run, loser.join)
-          }
-          result.flatMap {
-            case (exit, loser) =>
-              exit.foldM(
-                Cause.sequenceCauseOption(_) match {
-                  case Some(e) =>
-                    loser.merge.interrupt.as(Exit.halt(e.map(Some(_))))
-                  case None =>
-                    loser.fold(
-                      termination[E, O](leftDone, true, l),
-                      termination[E1, O2](true, rightDone, r)
-                    )
-                },
-                chunk => ZIO.succeedNow(Exit.succeed((chunk, (leftDone, rightDone, Some(loser)))))
-              )
-          }
+      for {
+        queue   <- Queue.unbounded[Take[E1, O3]].toManaged(_.shutdown)
+        done    <- RefM.make[Option[Boolean]](None).toManaged_
+        chunksL <- self.process
+        chunksR <- that.process
+        handler = (pull: Pull[R1, E1, O3], terminate: Boolean) => {
+          pull
+            .foldCause(c => Left(Cause.sequenceCauseOption(c)), Right(_))
```

Comment:
It looks like you could avoid some allocation by just using `run` here and doing a fold over the result in the modify section below
```suggestion
            .run
```

Reply:
@regiskuckaertz done, thanks